### DISCUSSION TITLE
Discover which endpoints are associated with SignalR hubs

### DISCUSSION THREAD

**Discussion by Ninds** [dotnet/aspnetcore#23685] [General] [answered]
Hi,
Is it possible to discover at runtime which endpoints exposed by one's own kestrel service are actually endpoints associated with SignalR hubs ?

### Context
The setup is described at davidfowl/BedrockFramework#65
_To summarize :_
The service is setup to handle multiple protocols of which one is HTTP.  The service is also setup with a (main) Controller AssemblyLoadContext and some worker AssemblyLoadContexts.  The controller ALC encapsulates an instance of Kestrel  wired up to a socket. The worker ALCs each encapsulate an instant of Kestrel listening on a MemoryTransport.  
Upon receiving requests, the Controller Kestrel dispatches (most but not all) requests internally to one of the worker Kestrels hosted in their own isolated ALCs. The main purpose/motivation for this choice of design is that the code servicing the requests cannot be guaranteed to be thread-safe and hence this particular implementation of isolation/sandboxing has been chosen.  
This setup works well in that it meets it's requirements and expectations.

I would however like to extend this framework to facilitate SignalR. I think it would be more appropriate for the hubs to hosted inside the controller Kestrel (or at least in a dedicated ALC/Kestrel per hub type) . To then dispatch a request to a worker Kestrel/ALC or process it in the controller Kestrel, the code would have to analyse whether the incoming request is headed for a SignalR hub or not.

Thanks
N

**Comment by Ninds** [comment, marked as answer]
I think:
[EndpointDataSource](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.routing.endpointdatasource?view=aspnetcore-3.1) 
[Endpoint.Metadata](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.endpoint.metadata?view=aspnetcore-3.1#Microsoft_AspNetCore_Http_Endpoint_Metadata) 
[HubMetadata](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.hubmetadata?view=aspnetcore-3.1) 

Should do the trick.

**Reply by BrennanConroy** [reply]
Haha, I was just about to answer with some of those links.

You'll also probably want to look for `NegotiateMetadata` as well since it is for the /hub/negotiate endpoint
https://github.com/dotnet/aspnetcore/blob/4b796c7bad1aa5f5aab6980ffe4f3ba78d4318e6/src/SignalR/common/Http.Connections/src/ConnectionEndpointRouteBuilderExtensions.cs#L101

**Reply by Ninds** [reply]
Thanks.  It's working nicely. 

> 

> ...probably want to look for NegotiateMetadata as well.. 

It wasn't necessary since `HubMetadata` covered the negotiate ones too.
